src/librustc_privacy/lib.rs

```diff
@@ -951,18 +949,10 @@ impl<'a, 'tcx: 'a> TypeVisitor<'tcx> for SearchInterfaceForPrivateItemsVisitor<'
                     self.min_visibility = vis;
                 }
                 if !vis.is_at_least(self.required_visibility, self.tcx) {
-                    if self.tcx.sess.features.borrow().pub_restricted || self.has_old_errors {
-                        let mut err = struct_span_err!(self.tcx.sess, self.span, E0446,
-                            "private type `{}` in public interface", ty);
-                        err.span_label(self.span, &format!("can't leak private type"));
-                        err.emit();
-                    } else {
-                        self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
-                                               node_id,
-                                               self.span,
-                                               format!("private type `{}` in public \
-                                                        interface (error E0446)", ty));
-                    }
+                    let mut err = struct_span_err!(self.tcx.sess, self.span, E0446,
```

Comment:
`private_in_public` can't be turned into an error in the near future unfortunately.
`self.tcx.sess.features.borrow().pub_restricted` can be well approximated with `is_pub_restricted(vis) && is_pub_restricted(self.required_visibility)` where `is_pub_restricted = not pub && not private`.
The idea is to always report an error and not warning when `pub(something)` is involved.